core/common/src/main/java/alluxio/network/thrift/BootstrapServerTransport.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.network.thrift;
+
+import static alluxio.network.thrift.BootstrapClientTransport.BOOTSTRAP_HEADER;
+
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.apache.thrift.transport.TTransportFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+/**
+ * The server side of bootstrap transport. This transport can fallback to the real RPC transport
+ * if the magic number is not found in the message header.
+ */
+public class BootstrapServerTransport extends TTransport {
+  private static final Logger LOG = LoggerFactory.getLogger(BootstrapServerTransport.class);
+
+  /** The base transport which we can peek into. */
+  private final PeekableTransport mBaseTransport;
+  /** The logic transport to work on, can be base transport or the real transport. */
+  private TTransport mTransport;
+  /** The factory to create the logic transport on open. */
+  private final TTransportFactory mTransportFactory;
+
+  /**
+   * @param baseTransport base transport
+   * @param tf transport factory to create the fallback transport
+   */
+  public BootstrapServerTransport(TTransport baseTransport, TTransportFactory tf) {
+    mBaseTransport = new PeekableTransport(baseTransport);
+    mTransportFactory = tf;
+  }
+
+  @Override
+  public void open() throws TTransportException {
+    LOG.debug("opening server transport");
+    if (!mBaseTransport.isOpen()) {
+      mBaseTransport.open();
+    }
+    byte[] messageHeader = new byte[BOOTSTRAP_HEADER.length];
+    int bytes;
+    try {
+      bytes = mBaseTransport.peek(messageHeader, 0, BOOTSTRAP_HEADER.length);
+    } catch (TTransportException e) {
+      if (e.getType() == TTransportException.END_OF_FILE) {
+        LOG.debug("No data in the stream {}", mBaseTransport);
+        mBaseTransport.close();
+        throw new TTransportException("No data data in the stream.");
+      }
+      throw e;
+    }
+
+    if (bytes == BOOTSTRAP_HEADER.length && Arrays.equals(messageHeader, BOOTSTRAP_HEADER)) {
+      mBaseTransport.consumeBuffer(BOOTSTRAP_HEADER.length);
+      mTransport = mBaseTransport;
+    } else {
+      mTransport = mTransportFactory.getTransport(mBaseTransport);
+    }
+    if (!mTransport.isOpen()) {
+      mTransport.open();
+    }
+  }
+
+  @Override
+  public boolean isOpen() {
+    return mBaseTransport.isOpen() && mTransport != null && mTransport.isOpen();
+  }
+
+  @Override
+  public void close() {
+    if (isOpen()) {
+      mTransport.close();
+    }
+  }
+
+  @Override
+  public int read(byte[] buf, int off, int len) throws TTransportException {
+    if (!isOpen()) {
+      throw new TTransportException("transport is not open");
```

Comment:
For this message (and the others like it), can we distinguish between the 2 transports?